[Compose 페이먼츠 2단계] 비비 미션 제출합니다.

rosemin928 opened the pull request:
# 셀프 체크리스트

 <!--
 모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
 -->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

 <!--
 학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
 정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
 -->
 
안녕하세요 제리~!
언제나 꼼꼼한 리뷰 감사합니당 덕분에 더 열심히 공부하고 있어요 ㅎㅎ
이번에도 잘부탁드립니다!! 👍

- 이번에 카드 갯수에 따른 화면을 구현하기 위해 `PaymentCardsContent`를 만들고 내부에서 `EmptyCard`, `SingleCard`, `MultiCard`로 바꾸면서 보여주도록 했습니다! 이 부분을 제리는 어떻게 구현하셨는지 궁금합니다!
- 저번에 리뷰해주신 리컴포지션에 따른 객체의 재생성 부분은 [refactor: 리컴포지션에 따른 객체 재생성 방지](https://github.com/woowacourse/android-payments/commit/06471beddea61425d84d01bf7da17fed768f5867) 해당 커밋을 참고해주세용

## 스크린샷
 <!-- 필요하다면 스크린샷을 첨부하세요. -->
https://github.com/user-attachments/assets/88862357-4439-421c-b4f4-3210bd93e341

rosemin928 committed:
docs: 2단계 기능명세서 작성

rosemin928 committed:
refactor: 패키지 이동

rosemin928 committed:
feat: 카드 목록이 비어있을 때 카드 목록 구현

rosemin928 committed:
feat: 카드 추가 기능 구현

rosemin928 committed:
feat: 카드 1장 있을 때의 화면 구현

rosemin928 committed:
feat: 카드 2장 이상 있을 때의 화면 구현

rosemin928 committed:
refactor: 카드 번호 포맷팅 적용

rosemin928 committed:
feat: 카드가 등록 후에 토스트를 노출하는 기능 구현

rosemin928 committed:
refactor: PaymentCardsContent 분리 및 내부에 들어갈 컴포넌트 선언

rosemin928 committed:
refactor: 리컴포지션에 따른 객체 재생성 방지

rosemin928 committed:
test: PaymentCardsContent에 대한 테스트 추가

rosemin928 committed:
test: PaymentCard에 대한 테스트 추가

rosemin928 committed:
test: PaymentCardsTopBar에 대한 테스트 추가

rosemin928 committed:
refactor: private 함수로 변경

jerry8282 commented on app/src/main/java/woowacourse/payments/ui/AddPaymentCardScreen.kt:
```diff
@@ -36,8 +47,14 @@ fun AddPaymentCardScreen() {
                 modifier = Modifier.padding(bottom = 14.dp),
-                onBackClick = { },
-                onSaveClick = { },
+                onBackClick = { (context as? Activity)?.finish() },
```
Activity를 finish()하는 것에 대한 책임은 누구에게 있는 게 좋을까요?🤔

rosemin928 replied:
Activity의 책임인 것 같네요!
Composable은 저장이 성공했는지의 여부만 알리고 성공 여부에 따라 화면을 닫을지는 Activity가 판단하도록 하는게 좋을 것 같아용

[수정했습니다!](https://github.com/woowacourse/android-payments/pull/38/commits/1652f4be00aee1a93a6e3d97c3c890571ef1691a)

jerry8282 commented on app/src/main/java/woowacourse/payments/ui/AddPaymentCardScreen.kt:
```diff
@@ -36,8 +47,14 @@ fun AddPaymentCardScreen() {
+                    val paymentCard = PaymentCard(cardNumber, expiry, owner, pin).toUiModel()
+                    (context as? Activity)?.apply {
+                        setResult(Activity.RESULT_OK, Intent().putExtra(EXTRA_CARD, paymentCard))
+                        finish()
```
setResult를 하는 것은 Screen의 책임일까요?🤔

rosemin928 replied:
이 또한 Activity의 책임인 것 같네요 🥲

[해당 부분도 수정했습니다!](https://github.com/woowacourse/android-payments/pull/38/commits/1652f4be00aee1a93a6e3d97c3c890571ef1691a)

jerry8282 commented on app/src/main/java/woowacourse/payments/domain/PaymentCard.kt:
```diff
@@ -0,0 +1,8 @@
+    val expiry: String,
+    val owner: String,
+    val pin: String,
+)
```
이 domain의 역할은 무엇인가요?🤔

rosemin928 replied:
나중에 서버로 보내게 되거나, 추가적으로 도메인 로직이 생길 수도 있다고 생각해서 도메인을 먼저 만들어두었습니다!

또, 현재 카드 화면에서는 pin을 보여주고 있지 않기 때문에 UiModel에는 pin을 제외했는데, 이 점이 도메인과 차이가 있다고 생각해서 도메인을 따로 만들기도 했습니다!

이번에 추가적으로 유효성 검증 로직을 구현하면서 여러 코드들이 추가되었는데 이에 대한 제리의 의견이 궁금합니다!

[feat: 카드 정보 유효성 검사 기능 추가](https://github.com/woowacourse/android-payments/pull/38/commits/8ac9e6795f4ca53c1bce469a9aab65e42afddca7)

jerry8282 commented on app/src/main/java/woowacourse/payments/ui/PaymentCardsScreen.kt:
```diff
@@ -0,0 +1,70 @@
+    fun launchAddCard() {
+        val intent = AddPaymentCardActivity.newIntent(context)
+        cardAddLauncher.launch(intent)
+    }
```
이렇게 하는 것과 파일 내에 private으로 최상위에 두는 것과 어떤 차이가 있나요?🤔

jerry8282 replied:
또한, cardAddLauncher를 통해 다른 Activity로 넘어가는 건 누구의 책임인지 비비의 생각이 궁금합니다.😊

rosemin928 replied:
`launchAddCard()`가 context와 cardAddLauncher를 사용하고 있는데, 둘 다 PaymentCardsScreen의 local scope 안에서만 유효하기 때문에 로컬 함수로 적어주었습니다.

이번에 해당 로직 전체를 Activity로 옮기며,` launchAddCard()`를 최상위 private 함수로 빼고, cardAddLauncher를 `private lateinit var`로 빼보려고 했습니다. 

하지만 이렇게 작성하게 되면 메모리 누수가 생기거나 구성 변경에 취약해지기 때문에 최종적으로는 onCardAdded라는 콜백을 연결해서 launcher 자체는 Activity의 생명주기에 1회만 등록하고, 상태 변경을 전달하는 것은 콜백이 담당하도록 수정해보았습니다!

[refactor: 카드 추가 로직을 Activity로 이동](https://github.com/woowacourse/android-payments/pull/38/commits/cb36e671500aa68e3c7aab4d4751c8fe27d9ad4e)

jerry8282 commented on app/src/main/java/woowacourse/payments/ui/format/CardFormatter.kt:
CardFormatter에 있는 메서드들은 PaymentCardUiModel에 있어도 될 거 같네요~

rosemin928 replied:
수정했습니다~!

[refactor: CardFormatter 클래스 제거 및 관련 함수 PaymentCardUiModel로 이동](https://github.com/woowacourse/android-payments/pull/38/commits/4775e4b1dd19bb0fb5e4916017eae7d3a1c80356)

jerry8282 commented on app/src/main/java/woowacourse/payments/ui/format/CardFormatter.kt:
```diff
@@ -0,0 +1,13 @@
+    if (raw.length <= 8) return raw
+    val masked = raw.take(8) + "*".repeat(raw.length - 8)
+    return masked.chunked(4).joinToString(separator)
+}
```
이 메서드는 특정 뷰에 종속적이게 될 거 같네유🥲

rosemin928 replied:
이 메서드가 특정 뷰에 종속적이게 될 것 같다는게 어떤 의미로 말씀하신건가요?? 😢

jerry8282 replied:
예를 들어 CardNumber의 끝에 4자리만 가리게 하는 뷰가 생기거나 가리는 mask가 "*"이 아닌 "•"로 표현되는 뷰가 생긴다면 이 메서드는 하나의 뷰에서만 쓰이게 될 거 같다는 말이였습니다.😁

jerry8282 commented on app/src/main/java/woowacourse/payments/ui/model/PaymentCardUiModel.kt:
```diff
@@ -0,0 +1,13 @@
+    val owner: String,
+) : Parcelable
+
+const val EXTRA_CARD = "extra_card"
```
intent를 만들기위한 KEY값이 UiModel에 있어도 괜찮나요?🤔

rosemin928 replied:
허걱 아니요...Activity로 옮겨주었습니다!

jerry8282 commented on app/src/main/java/woowacourse/payments/ui/component/PaymentCardsTopBar.kt:
```diff
@@ -0,0 +1,53 @@
+    CenterAlignedTopAppBar(
+        title = { Text(stringResource(R.string.title_payment_cards)) },
+        actions = {
+            if (cardCount >= 2) {
```
TopBar에서 cardCount를 알 필요가 있을까유?🤔

rosemin928 replied:
단순하게 "카드 추가 버튼"이 보여야 하는지, 보이지 않아야 하는지만 알면 될 것 같네요 😅

[수정했습니다!](https://github.com/woowacourse/android-payments/pull/38/commits/d4399f4d83970735b446e124acf643ee9ae883ce)

jerry8282 commented on app/src/main/res/values/strings.xml:
```diff
@@ -16,4 +16,11 @@
     <string name="card_number_separator">&#160;-&#160;</string>
     <string name="expiry_separator">&#160;/&#160;</string>
```
구분자를 const 로 빼지 않고 string.xml로 뺀 이유가 무엇인가요?🤔

rosemin928 replied:
저는 화면에 보여지는 문자들을 string.xml로 빼고, 화면에 필요한 상수나 문자열들은 const로 빼는 편입니다!

구분자는 화면에서 보여지는 것들이기 때문에 string.xml로 빼주었는데, 제리는 string.xml로 빼는 것과 const로 빼는 것에 대한 기준이 있으신가요??

jerry8282 replied:
제가 생각하기에 string.xml로 분리하는 이유는 다국어 지원을 위함이라고 보고 있습니다. 
그래서 보통 언어와 관련된 문자열들은 외부로 빼주는 편인데요. 

그렇다면 비비님 의견으로는 마스킹에 사용되는 * 기호도 string.xml로 분리하는 것이 맞다고 생각하시는지 궁금합니다.😁

rosemin928 replied:
음...저는 이때까지는 마스킹 기호도 분리하는게 맞다고 생각을 했는데(그래놓고 분리하는걸 빼먹었네요 ㅋㅋㅋ 😅) 다국어 지원을 위해 string으로 분리한다는 제리의 말도 맞는 것 같네요! 보통 다국어를 위한 string들을 빼는 것 같기도 하구요!

이 부분은 3단계 구현을 하면서 정리해보도록 하겠습니다! 🙇‍♀️

jerry8282 commented on app/src/main/java/woowacourse/payments/ui/format/CardFormatter.kt:
```diff
@@ -0,0 +1,13 @@
+package woowacourse.payments.ui.format
+
+fun String.maskedCardNumber(separator: String): String {
+    val raw = filter { it.isDigit() }
```
`val raw = filter { it.isDigit() }`
이 코드는 없어도 잘 작동이 될 거 같은데 넣으신 의도가 무엇인가요?🤔

rosemin928 replied:
도메인에서 한 번 더 검증을 하는게 좋지 않을까?하는 생각에 숫자만 필터링하는 로직을 추가했는데, 현재 이미 숫자만 입력하도록 하고 있고 키보드 자체도 숫자 키보드로 바꿔주었기 때문에 괜찮을 것 같아서 해당 코드는 삭제했습니다!

jerry8282 commented on app/src/main/java/woowacourse/payments/ui/PaymentCardsScreen.kt:
```diff
@@ -0,0 +1,70 @@
+fun PaymentCardsScreen() {
+    val context = LocalContext.current
+    val toastMessage = stringResource(R.string.toast_card_add)
+    var paymentCards by rememberSaveable { mutableStateOf(listOf<PaymentCardUiModel>()) }
```
paymentCards를 Screen 내부에 상태로 들고 있으면 Screen에 대한 테스트나 Preview를 보기가 힘들어질 거 같아요🥲

rosemin928 replied:
paymentCards를 Activity로 빼주었습니다!

[수정했습니다~](https://github.com/woowacourse/android-payments/pull/38/commits/cb36e671500aa68e3c7aab4d4751c8fe27d9ad4e)

jerry8282 commented on app/src/main/java/woowacourse/payments/ui/PaymentCardsScreen.kt:
```diff
@@ -0,0 +1,70 @@
+                val newPaymentCard = result.data?.getParcelableExtra<PaymentCardUiModel>(EXTRA_CARD)
+                if (newPaymentCard != null) {
+                    paymentCards = paymentCards + newPaymentCard
+                    Toast.makeText(context, toastMessage, Toast.LENGTH_SHORT).show()
```
```suggestion
                    Toast.makeText(context, context.getString(R.string.toast_card_add), Toast.LENGTH_SHORT).show()
```
위에서 받아온 context를 쓰지 않고 toastMessage라고 따로 빼둔 이유가 무엇인가요?🤔

rosemin928 replied:
stringResource가 Composable 함수이기 때문에 리컴포지션 될 때마다 현재 Locale이나 Configuration change를 반영해서 최신 문자열을 가져옵니다.

예를 들어서 앱 실행 도중에 시스템 언어를 바꾸면, 토스트 메시지 값도 리컴포지션 시점에 새 문자열로 업데이트가 되기 때문에 toastMessage라고 따로 상태처럼 기억해두고 쓰도록 했습니다!

하지만 이번에 제리의 리뷰를 참고해서 해당 로직들을 전부 Activity로 옮기며 토스트 메시지 자체가 리컴포지션과 무관해졌기 때문에 다음처럼 수정했습니다!

[refactor: 카드 추가 로직을 Activity로 이동](https://github.com/woowacourse/android-payments/pull/38/commits/cb36e671500aa68e3c7aab4d4751c8fe27d9ad4e)

jerry8282 commented on app/src/main/java/woowacourse/payments/ui/AddPaymentCardScreen.kt:
```diff
@@ -72,7 +89,7 @@ fun AddPaymentCardScreen() {
-                    NumberVisualTransformation(
+                    RememberNumberVisualTransformation(
                         2,
                         stringResource(R.string.expiry_separator),
```
NumberTextField는 파라미터로 Int를 받은 뒤에 Res값을 넣어주시고 여기에는 stringResource를 넣어서 String 값을 넣어주셨는데 둘의 차이가 무엇인가요?🤔

rosemin928 replied:
NumberTextfield는 stringResource를 직접 호출해서 Compose를 의존하고, VisualTransformation은 Composable이 아닌 순수 UI 유틸이라 Compose에 직접 의존하지 않도록 하기 위해 String으로 넣어주었습니다...!

jerry8282 commented on app/src/main/java/woowacourse/payments/ui/component/PaymentCardsContent.kt:
```diff
@@ -0,0 +1,114 @@
+    paymentCards: List<PaymentCardUiModel>,
+    onAddCard: () -> Unit,
+) {
+    when (paymentCards.size) {
+        0 -> EmptyCard(modifier = modifier, onAddCard = onAddCard)
+        1 ->
+            SingleCard(
+                modifier = modifier,
+                paymentCard = paymentCards.first(),
+                onAddCard = onAddCard,
+            )
+
+        else -> MultiCards(modifier = modifier, paymentCards = paymentCards)
```
저는 카드 개수를 단순히 비교해서 컴포넌트를 추가하는 방식으로 구현했었는데, 상태별로 Composable을 분리해 두신 걸 보니 훨씬 가독성이 좋아 보이네요. 덕분에 많이 배워갑니다!😊

jerry8282 commented on app/src/main/java/woowacourse/payments/ui/theme/Type.kt:
```diff
@@ -4,31 +4,26 @@ import androidx.compose.material3.Typography
+val CardTextStyle =
+    TextStyle(
+        fontSize = 12.sp,
+        lineHeight = 12.sp,
+        letterSpacing = 0.12.em,
         fontWeight = FontWeight.Medium,
-        fontSize = 11.sp,
-        lineHeight = 16.sp,
-        letterSpacing = 0.5.sp
+        fontFamily = Roboto,
     )
```
textStyle을 공통으로 묶어주셨군요! 덕분에 중복이 줄어들고 코드도 훨씬 깔끔해진 것 같아요 🙌

jerry8282 reviewed (changes requested):
비비 안녕하세요~

리뷰가 늦어진 점 죄송합니다.🥲
대신 더더욱 꼼꼼히 리뷰를 하였습니다!
비비가 리뷰 반영과 생각 정리를 너무 잘해주셔서 리뷰가 즐겁네요😊

 2단계 미션 구현 고생 많으셨습니다!  이번에도 미션 구현을 너무나 잘해주셨습니다 💯
 비비의 코드를 보면서 궁금한 점들과 추가로 고민해 보시면 좋을 만한 
내용들에 코멘트를 달았으니 확인해주시면 감사하겠습니다~

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏

rosemin928 committed:
refactor: 화면 간 데이터 전달 방식 변경

rosemin928 committed:
refactor: CardFormatter 클래스 제거 및 관련 함수 PaymentCardUiModel로 이동

rosemin928 committed:
feat: 카드 정보 유효성 검사 기능 추가

- 카드 번호, 만료일, PIN에 대한 유효성 검사 로직을 도메인 모델에 추가
- 유효하지 않은 입력 값에 대한 예외 처리 추가
- 유효성 검사 결과에 따라 UI에 오류 메시지 표시
- `NumberTextField`에 오류 상태 및 보조 텍스트 표시 기능 추가

rosemin928 committed:
refactor: 카드 추가 버튼 노출 조건 변경

rosemin928 committed:
refactor: 카드 추가 로직을 Activity로 이동

rosemin928 committed:
refactor: 숫자만 필터링하는 로직 제거 및 상수화

rosemin928 committed:
refactor: InvalidCardException으로 예외 통합

rosemin928 reviewed (commented):
안녕하세요 제리!
꼼꼼한 리뷰 감사합니다 😊 시간이 좀 많이 걸렸네요...ㅠㅠ

이번에 유효성 검증을 추가적으로 작성했는데 이에 대한 제리의 의견이 궁금합니다!
또, stateHolder를 추가해봤는데 이렇게 사용하는게 맞는지 잘 모르겠어서 이에 대해서도 제리의 의견이 궁금합니다!
이 부분에 대해 집중적으로 리뷰를 해주시면 정말 감사할 것 같습니다!! 🙇‍♀️

+) 스크린샷을 업데이트 해두었으니 참고해주세용

rosemin928 committed:
refactor: 중복 검증 삭제

rosemin928 committed:
refactor: 숫자 입력 컴포넌트 분리

- `NumberTextField`를 `CardNumberTextField`, `ExpiryTextField`, `PinTextField`로 분리
- `StringTextField`를 `OwnerTextField`로 이름 변경

rosemin928 committed:
test: 숫자 컴포넌트 분리에 따른 테스트 수정

rosemin928 committed:
refactor: 패키지 이동

rosemin928 committed:
feat: AddPaymentCardStateHolder 클래스 및 AddPaymentCardUiState 데이터 클래스 생성

rosemin928 committed:
refactor: `NewCardTopBar` 저장 버튼 활성화 상태 추가

jerry8282 commented on app/src/main/java/woowacourse/payments/domain/model/CardNumber.kt:
```diff
@@ -0,0 +1,16 @@
+import woowacourse.payments.domain.exception.InvalidCardException
+
+@JvmInline
+value class CardNumber private constructor(
+    val value: String,
+) {
+    companion object {
+        private const val CARD_NUMBER_LENGTH = 16
+
+        fun from(raw: String): CardNumber? = if (raw.length == CARD_NUMBER_LENGTH) CardNumber(raw) else null
+
+        fun require(raw: String): CardNumber = from(raw) ?: throw InvalidCardException.InvalidCardNumber()
+    }
+}
```
```suggestion
value class CardNumber(
    val value: String,
) {
    init {
        if (value.length != CARD_NUMBER_LENGTH) {
            throw InvalidCardException.InvalidCardNumber()
        }
    }

    companion object {
        private const val CARD_NUMBER_LENGTH = 16
    }
}

```

이 코드와 비교해서 비비가 작성해주신 코드는 어떠한 장점이 있나요?🧐

rosemin928 replied:
제가 작성한 코드에서는 외부에서는 `CardNumber("~~")`와 같이 직접 생성이 불가하고, 반드시 `from` 혹은 `require`로만 생성을 하도록 했습니다. 이렇게 무조건 팩토리를 거치게 하므로 유효성 로직을 한 곳에서 통제할 수 있다는 장점이 있습니다!

저는 특히 `CardNumber("~~")`를 아예 막는 점이 좋다고 생각했는데 제리는 init에서 유효성 검증을 하는 것이 더 좋다고 생각하시나요?? 제리의 의견이 궁금합니다!

jerry8282 commented on app/src/main/java/woowacourse/payments/ui/add/AddPaymentCardStateHolder.kt:
```diff
@@ -0,0 +1,63 @@
+import woowacourse.payments.ui.model.mapper.toUiModel
+
+class AddPaymentCardStateHolder private constructor(
+    private val stateHolder: MutableState<AddPaymentCardUiState>,
```
nit: stateHolder란 네이밍 보단 uistate란 네이밍은 어떠신가용?🤣

rosemin928 replied:
`AddPaymentCardUiState`가 따로 있어서 stateHolder라는 이름을 쓰게 되었어요 🥲
그래서 만약 바꾸게 된다면 `AddPaymentCardState`, `AddPaymentCardController` 같은 이름이 될 것 같은데 이런 이름들보다는 stateHolder라는 이름이 더 나을 것 같은데 제리는 어떻게 생각하시나요??

jerry8282 replied:
> private val stateHolder: MutableState AddPaymentCardUiState

이 프로퍼티의 변수명을 말한 겁니다유 AddPaymentCardUiState인데 uistate란 변수명이 더 맞지 않을까 하는 생각

jerry8282 commented on app/src/main/java/woowacourse/payments/ui/component/PaymentCardsContent.kt:
```diff
@@ -0,0 +1,114 @@
+    }
+}
+
+@Preview(showBackground = true)
+@Composable
+private fun EmptyCardPreview() {
+    EmptyCard(onAddCard = {})
+}
+
+@Preview(showBackground = true)
+@Composable
+private fun SingleCardPreview() {
+    SingleCard(
+        onAddCard = {},
+        paymentCard = PaymentCardUiModel("1234567812345678", "0511", "minjeong"),
+    )
+}
+
+@Preview(showBackground = true)
+@Composable
+private fun MultiCardsPreview() {
+    MultiCards(
+        paymentCards =
+            listOf(
+                PaymentCardUiModel("1234123456785678", "1215", "minjeong"),
+                PaymentCardUiModel("1111222233334444", "1234", "junseo"),
+            ),
+    )
+}
```
Screen에서 preview를 하지 않고 ScreenContent에서 preview를 한 이유가 무엇인가요?🧐

jerry8282 replied:
추가로, PreviewParameterProvider를 안쓰신 이유가 있으신가요?🧐

rosemin928 replied:
오...`PreviewParameterProvider`는 몰라서 못 썼습니다 🤣
그런게 있군용 다음에 한 번 써보도록 하겠습니당

ScreenContent는 순수 UI이고, Screen에는 viewModel, StateHolder, Launcher 등이 들어가게 될 수도 있다고 생각해서 preview의 안정성을 위해 screenContent에서 preview를 했습니다~!

jerry8282 commented on app/src/main/java/woowacourse/payments/ui/cards/PaymentCardsScreen.kt:
```diff
@@ -0,0 +1,43 @@
+
+@Composable
+fun PaymentCardsScreen(
+    paymentCards: List<PaymentCardUiModel>,
```
앞에 스크린은 StateHolder를 사용하셨는데 여기도 UiModel를 직접 넣는 거 보다 StateHolder를 사용하면 더 깔끔할 거 같긴 하네요.🤣

rosemin928 replied:
그렇네용 StateHolder 사용해서 수정해주었습니다!

[수정했습니다~~](https://github.com/woowacourse/android-payments/pull/38/commits/c6600cbfcce2bed1502ba183b325c6733a3a0f90)

jerry8282 commented on app/src/main/java/woowacourse/payments/ui/add/AddPaymentCardStateHolder.kt:
```diff
@@ -0,0 +1,63 @@
+    fun buildResult(): Result<PaymentCardUiModel> =
+        PaymentCard
+            .create(state.cardNumber, state.expiry, state.owner, state.pin)
+            .map { it.toUiModel() }
```
```suggestion
fun buildResult(): Result<PaymentCardUiModel> =
        runCatching {
            PaymentCard
                .create(state.cardNumber, state.expiry, state.owner, state.pin)
                .toUiModel()
        }
```
domain에서 Result객체를 반환해도 괜찮나요?🧐

rosemin928 replied:
성공했을 때/실패했을 때를 나눠야할 것 같아서 Result를 썼었는데 지금 보니 실패했을 때를 아예 작성하지 않기도 했고, 사용하는 위치도 적절하지 않은 것 같네요 😅

[수정했습니다!](https://github.com/woowacourse/android-payments/pull/38/commits/7dbf1c0a05e6f6ffd4317d59178902168ad72d33)

jerry8282 commented on app/src/main/java/woowacourse/payments/ui/cards/PaymentCardsActivity.kt:
```diff
@@ -0,0 +1,67 @@
+
+class PaymentCardsActivity : ComponentActivity() {
+    private lateinit var cardAddLauncher: ActivityResultLauncher<Intent>
+    private var onCardAdded: ((PaymentCardUiModel) -> Unit)? = null
```
두번째 화면에서 카드 내용들을 다 입력하고 화면을 회전한 뒤에 카드를 추가하면 어떤 일이 발생할까요?🧐

rosemin928 replied:
오...카드 추가가 안 되는군용...꼼꼼한 리뷰 감사합니다 🤣

[수정했습니당](https://github.com/woowacourse/android-payments/pull/38/commits/a1eb5158d9b28e5ea8b6e5de629a425ad19a5a8e)

jerry8282 commented on app/src/main/java/woowacourse/payments/ui/component/CardNumberTextField.kt:
```diff
@@ -0,0 +1,57 @@
+    onValueChange: (String) -> Unit,
+    modifier: Modifier = Modifier,
+) {
+    val separator = stringResource(R.string.card_number_separator)
+    val visualTransformation = remember(separator) { NumberVisualTransformation(4, separator) }
+    val textState = remember(value) { TextFieldValue(value, selection = TextRange(value.length)) }
+
+    val digitsOnly = value.filter(Char::isDigit)
+    val showError = digitsOnly.isNotEmpty() && CardNumber.from(digitsOnly) == null
```
이부분이 상태로 있으면 재사용성이 많이 떨어질 것 같네요🥲
비비는 컴포즈를 만들 때 어디까지 인자로 빼야 된다고 생각하시나요?🧐

rosemin928 replied:
원래는 `NumberTextField`로 재사용을 할 수 있도록 컴포넌트들을 만들어주었었는데 만들다보니까 `NumberTextField`에 인자가 너무 많아지고 지나치게 뚱뚱해지는 것 같아서 그냥 전부 분리를 해주었습니다.
그래서 재사용을 아예 염두에 두지 않고 만들었는데 그러면 그냥 가벼운게 낫지 않나? 싶어서 많은 것들을 내부에서 상태로 관리하도록 작성했습니다!

어디까지 인자로 뻬야할지는 상황에 따라 다르게 정할 것 같습니다 ㅎㅎ... 만약 재사용을 하고 싶다면 필요한 부분을 인자로 둘 것 같습니당

jerry8282 commented on app/src/main/java/woowacourse/payments/ui/add/AddPaymentCardStateHolder.kt:
```diff
@@ -0,0 +1,63 @@
+
+    fun onPinChange(new: String) = update { it.copy(pin = new) }
+
+    val isCardNumberValid: Boolean
+        get() = state.cardNumber.isNotEmpty() && CardNumber.from(state.cardNumber) != null
+    val isExpiryValid: Boolean
+        get() = state.expiry.isNotEmpty() && Expiry.from(state.expiry) != null
+    val isPinValid: Boolean
+        get() = state.pin.isNotEmpty() && Pin.from(state.pin) != null
```
현재 상태에서는 사용자가 한 단어만 입력해도 에러 문구가 뜨도록 되어 있는데 의도하신 부분인가용?🧐

rosemin928 replied:
네! 16자리/4자리 숫자를 전부 다 입력하기 전까지는 에러 문구가 떠있는 것이 더 알아보기 쉽다고 생각해서 의도해서 작성했습니당

jerry8282 commented on app/src/main/java/woowacourse/payments/ui/add/AddPaymentCardStateHolder.kt:
```diff
@@ -0,0 +1,63 @@
+        get() = state.pin.isNotEmpty() && Pin.from(state.pin) != null
+
+    val canSave: Boolean
+        get() = isCardNumberValid && isExpiryValid && isPinValid
```
현재 구조는 owner가 입력되어도 canSave가 계속해서 실행될 것 같네요🥲

rosemin928 replied:
허걱 그렇네요...수정했습니다!

[수정~~~](https://github.com/woowacourse/android-payments/pull/38/commits/1cf298b7baff08d4276af61cef315a2528c1e60a)

jerry8282 reviewed (changes requested):
비비 안녕하세요~
해당 미션 너무 고생 많으셨습니다!
리뷰 반영 정말 잘 해주셨네요.💯

이번 미션 너무 잘해주셨습니다👍
수정해야할 부분이 조금 있어서 다시 리뷰 드립니다.
반영을 다 너무 잘해주셔서 다음 리뷰 때 바로 머지하도록 하겠습니다.🤣
추가로 생각해보면 좋을 것을 남겨두었으니 확인해주시면 감사하겠습니다.

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏

rosemin928 committed:
refactor: canSave를 derivedStateOf로 변경

rosemin928 committed:
refactor: 카드 추가 로직 수정

rosemin928 committed:
refactor: Result 타입 대신 nullable 타입 사용

rosemin928 committed:
feat: PaymentCardsUiState 및 PaymentCardsStateHolder 추가

rosemin928 reviewed (commented):
안녕하세요 제리~!
이번에도 꼼꼼한 리뷰 감사합니당
덕분에 큰 도움이 되었어요!!

제리도 3단계 하시느라 바쁘실텐데 화이팅이구 이번에도 잘부탁드립니다~ ㅎㅎ

jerry8282 reviewed (approved):
비비 안녕하세요~
해당 미션 너무 고생 많으셨습니다!
리뷰 반영 정말 잘 해주셨네요.💯

이번 미션 너무 잘해주셨습니다👍
제가 리뷰에 대한 반영을 완벽하게 해주셨네요😊
추가적인 코멘트는 3단계에 같이 남겨드리도록 하겠습니다.🤓

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏
3단계도 화이팅 입니다~
